Expand extraction support for remote archive files.

jdost opened the pull request:
This allows for better detection support and adds an optional parameter to the repository definition to allow for explicitly overriding the extraction handler.  This is to allow for both `.tgz` files implicitly as well as allow for overriding the detection logic when the full URL doesn't provide an extension by default (like URLs from `codeload.github.com` which end with the commit SHA).

FuegoFro commented on Sources/RepoToolsCore/RepoActions.swift:
```diff
@@ -407,7 +407,7 @@ public enum RepoActions {
+            } else if lowercasedFileName.hasSuffix("tar.gz") || lowercasedFileName.hasSuffix("tgz") {
                 return shell.command(CommandBinary.sh, arguments: [
                     "-c",
                     untarTransaction(
```
(GitHub won't let me comment on the line I actually care about, which is below) While we're here we might want to update the message below to ...`other than .zip, .tar.gz, or .tgz`

jdost replied:
Fixed: https://github.com/discordapp/PodToBUILD/pull/1/commits/ebeffa42837078a873fcf48ad0b01fc20aa01bb6#diff-8899a6f8f09176941c964bb6deee3f2cR440

FuegoFro reviewed (approved):
This seems fine on its own, and if we need support for explicitly specifying type we can do that in another PR 🙂

jdost commented on Sources/RepoToolsCore/RepoActions.swift:
```diff
@@ -399,15 +401,34 @@ public enum RepoActions {
         func extract() -> CommandOutput {
             let lowercasedFileName = fileName.lowercased()
-            if lowercasedFileName.hasSuffix("zip") {
+            if fetchOptions.fileHandler?.isEmpty == false {
```
I am not a swift expert, so there may be a more elegant way of doing this cascading logic...  But I wanted to treat any definition of this override as different with no falling back (that way if the developer does define an invalid override it doesn't quietly fail and carry forward to extension checking).

FuegoFro replied:
I would do something like this:

```swift
let resolvedExtractionHandler: String
if fetchOptions.fileHandler == "default" {
    let lowercasedFileName = fileName.lowercased()
    if lowercasedFileName.hasSuffix("zip") {
        resolvedExtractionHandler = "zip"
    } else if lowercasedFileName.hasSuffix("tar.gz") || lowercasedFileName.hasSuffix("tgz") {
        resolvedExtractionHandler = "tar"
    } else {
        fatalError("Cannot extract files other than .zip, .tar.gz, or .tgz")
    }
} else {
    resolvedExtractionHandler = fetchOptions.fileHandler
}

if resolvedExtractionHandler == "zip" {
    return shell.command(CommandBinary.sh, arguments: [
        "-c",
        unzipTransaction(
            rootDir: extractDir,
            fileName: escape(download)
        ),
    ])
} else if resolvedExtractionHandler == "tar" {
    return shell.command(CommandBinary.sh, arguments: [
        "-c",
        untarTransaction(
            rootDir: extractDir,
            fileName: escape(download)
        ),
    ])
}
fatalError("Should not be possible to get here.")
```

jdost replied:
So I broke the logic down about the same way this is done: https://github.com/discordapp/PodToBUILD/pull/1/commits/43c63479fa365a25e868078b4768303f85124ded#diff-8899a6f8f09176941c964bb6deee3f2cR404

jdost commented on Sources/RepoToolsCore/RepoActions.swift:
```diff
@@ -25,6 +25,7 @@ public struct FetchOptions {
     public let url: String
     public let trace: Bool
     public let subDir: String?
+    public let fileHandler: String?
```
Hrm, I changed the name of this in the bazel side to `extraction handler`, may want to switch it here too, thoughts?

FuegoFro replied:
Seems reasonable to keep it consistent. Any reason to not name this `extractionHandler` here?

jdost replied:
I will switch it, I had been using this name and didn't realize I had switched until I saw this and made the comment.  I will make the switch when updating the swift code with the other suggestions.

jdost replied:
Switched this to `extractionHandler` to keep it in line with the naming in other changes.

FuegoFro commented on BazelExtensions/workspace.bzl:
```diff
@@ -180,6 +183,7 @@ pod_repo_ = repository_rule(
         "enable_modules": attr.bool(default=True, mandatory=True),
         "generate_module_map": attr.bool(default=True, mandatory=True),
         "header_visibility": attr.string(),
+        "extraction_handler": attr.string(),
```
I think a better way to do this might be to have `attr.string(values=["default", "zip", "tar"], default="default")` (so Bazel does some of the checking for us) and then expand the docs below to explicitly include the `"default"` option. This would also involve changing other defaults from `""` to `"default"`.

FuegoFro commented on Sources/RepoToolsCore/RepoActions.swift:
```diff
@@ -416,7 +437,7 @@ public enum RepoActions {
                 ])
             }
-            fatalError("Cannot extract files other than .zip or .tar")
+            fatalError("Cannot extract files other than .zip, .tar, .tar.gz, or .tgz")
```
This isn't actually right, since `.tar` wouldn't be valid.

jdost replied:
*sigh* you are right, which is really sad considering that this originally said `.tar`...  I will add that suffix check to the list of matches for tarballs.

jdost replied:
I added `tar` support: https://github.com/discordapp/PodToBUILD/pull/1/commits/43c63479fa365a25e868078b4768303f85124ded#diff-8899a6f8f09176941c964bb6deee3f2cR411

FuegoFro commented on BazelExtensions/workspace.bzl:
```diff
@@ -258,6 +263,9 @@ def new_pod_repository(name,
          https://github.com/bazelbuild/bazel/pull/3712
+
+         extraction_handler: override for how the target file from `url` is
+         extracted as, can be either `tar` or `zip`.
```
Can you add `default` here and describe what that does? 🙂

FuegoFro commented on BazelExtensions/workspace.bzl:
```diff
@@ -41,7 +41,9 @@ def _fetch_remote_repo(repository_ctx, repo_tool_bin, target_name, url):
-        "true" if repository_ctx.attr.trace else "false"
+        "true" if repository_ctx.attr.trace else "false",
+        "--handler",
+        repository_ctx.attr.extraction_handler
```
I know they weren't here before, but let's leave some trailing commas here 🙂

FuegoFro commented on bin/update_pods.py:
```diff
@@ -347,6 +352,9 @@ def new_pod_repository(name,
          https://github.com/bazelbuild/bazel/pull/3712
+
+         extraction_handler: override for what archive type to treat the file as
+         when extracting it, supports `tar` and `zip`.
```
Same with docs for `default`

FuegoFro commented on BazelExtensions/workspace.bzl:
```diff
@@ -279,5 +287,6 @@ def new_pod_repository(name,
         generate_module_map=generate_module_map,
-        header_visibility=header_visibility
+        header_visibility=header_visibility,
+        extraction_handler=extraction_handler
```
(trailing comma plz 🙌 )

jdost replied:
Wasn't sure if skylark allowed it so I opted to just stick with the existing convention.

jdost replied:
So I thought about this a bit and am going to keep this and the above as is... My reasoning is that if this is going to be merged to the upstream master it should try and follow their conventions.

FuegoFro replied:
That's a really good point, good call

![good-idea](https://user-images.githubusercontent.com/463696/58209300-9e5a3d80-7c9b-11e9-83cb-4ec17d070fb7.gif)

FuegoFro reviewed (approved):
Nice, looks great!

FuegoFro commented:
Hm, actually I'm not sure we should necessarily merge this in. All of the commits in `discord_fixes` come from PRs that are open into the upstream PodToBUILD (the idea being that if/when those get merged in, we can just blow away and recreate that branch from the latest upstream). Once we get this into a good state, how would you feel about instead closing this PR and opening one for this branch into upstream master? Then we can separately merge this branch into `discord_fixes` and do another release. Or maybe this is just over thinking things? I mainly want to be sure that we try to upstream these changes so we (hopefully) eventually don't need to keep this fork around.

jdost committed:
Expand extraction logic support

* Adds in support for file extensions of `.tar` and `.tgz` for the tar
  extraction logic
* Adds an optional override parameter of `extraction_handler` that
  allows to define an explicit extraction method to be used and override
  the inferring code (this is especially useful for URLs that don't end
  with the expected extension).

jdost commented:
So I rebased and changed the target to `master` (it didn't change the delta, so same changeset).  I added the `default` value to docstrings and explain the fallback behavior in them.  I did not add the trailing commas as I am going to stick with the existing code style for when I open an upstream PR with these changes.

FuegoFro reviewed (commented):
Looks great! I can go ahead and merge this into our fixes branch and create another release 🙂

jdost committed:
Fixed naming mess up

jdost committed:
Fix test to include new struct property

jdost committed:
Safety cleanup

* Add explicit value for debug printing for additional safety
* Add missed trailing comma in argument definition

rahul-malik committed:
Update README.md

FuegoFro committed:
Update Apple/Swift/Bazel versions to allow building with Xcode 10.2 (#78)

* Update Apple/Swift/Bazel versions to allow building with Xcode 10.2

The goal with this is to update the dependencies to include  This updates the `rules_apple` dependency to latest in order to pull in https://github.com/bazelbuild/rules_swift/commit/19db162c614af9c9db354e14d68965c8eb58e348 and allow compiling with Xcode 10.2. This also switches to using the version of `rules_swift` included by `rules_apple`, rather than manually specifying it.

Finally, this upgrades the version of Bazel being used to be compatible with the newer versions of `rules_apple` and `rules_swift`.

This seems like it partially addresses https://github.com/pinterest/PodToBUILD/issues/74, but that also mentions other tasks like updating other projects and CI.

* Further upgrades/fixes for 0.25.2

- Upgrade example projects to use 0.25.2
- Fix example projects to assume they're being compiled for iOS
- Change the image used to test in Travis
- Upgrade Swift libraries to work with Swift 5

FuegoFro committed:
Use fresh temporary directories (#79)

The current implementation of `tmpdir` just returns the root of the
user's global temp dir (as returned by `NSTemporaryDirectory()`). This
changes that to create a unique subdirectory within it. This means that
each invocation of `tmpdir` will now return a new, empty directory.

This helps avoid issues with output from one repository affecting
another repository. In particular when unzipping an archive from GitHub
I was seeing issues where `unzip` was asking whether `.gitignore` should
be overwritten and immediately failing with an EOF. This should provide
better isolation between repositories.

While I'm here, this also changes a rouge tab to spaces in a Python file.

FuegoFro committed:
Suppress stderr when untarring (#77)

At least on macOS 10.14.4, this `tar` invocation prints each of the files it is unpacking to stderr. It seems that nothing is actually reading the stderr pipe from this subprocess, and it therefore deadlocks and never finishes when untarring a repo with many files. This can be reproduced with the latest stable PodToBUILD and both Bazel 0.24.1 and 0.25.2 via the following WORKSPACE file (and, say, running `bazel query @React//...`):

```bzl
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "rules_pods",
    urls = ["https://github.com/pinterest/PodToBUILD/releases/download/0.22.0-ee8466e/PodToBUILD.zip"],
)

load("@rules_pods//BazelExtensions:workspace.bzl", "new_pod_repository")

new_pod_repository(
    name = "React",
    url = "https://github.com/facebook/react-native/tarball/4105450c7e15d6b46f88e2dad8a92cb28c3e882b/react-native.tar.gz",
    strip_prefix = "facebook-react-native-4105450/",
)
```

This change updates the command to also redirect the `tar`'s stderr to `/dev/null`, thereby allowing PodToBUILD to untar repos with many files.

jdost committed:
Add new argument definition to README docs

jdost committed:
Fix Test failure due to bad interpolation logic

FuegoFro commented:
I believe a version of this has been merged into upstream (namly https://github.com/pinterest/PodToBUILD/pull/87) and this PR/branch is no longer being used in our `discord_fixes` branch, so we can probably just close it?

jdost commented:
:+1:
